### PR TITLE
[Search] Introduce Result Item Extensions ✨

### PR DESCRIPTION
**Pull request by camilaibs**
Thank you, @iamEAP, for all your help!

## Hey, I just made a Pull Request!

Closes: https://github.com/backstage/backstage/issues/14603

Introduces search result item extensions:
- [x] Create an extension function to be provided by plugins;
- [x] Create a component and hook for rendering extensions;
- [x] Export extensions for catalog, explore, tools, and default result item;
- [x] Use result item extensions on search pages and modals.

Extensions for search results let you customize components used to render search result items; It is possible to provide your own search result item extensions or use the ones provided by plugin packages:

Using the example below, you can provide an extension to be used as a default result item:

```tsx
// plugins/your-plugin/src/plugin.ts
import { createPlugin } from '@backstage/core-plugin-api';
import { createSearchResultListItem } from '@backstage/plugin-search-react';

const plugin = createPlugin({ id: 'YOUR_PLUGIN_ID' });

export const YourSearchResultListItem = plugin.provide(
  createSearchResultListItemExtension({ ⬅️ 
    name: 'YourSearchResultListItem',
    component: () => import('./components').then(m => m.YourSearchResultListItem),
  }),
);
```

Additionally, you can define a predicate function that receives a result and returns whether your extension should be used to render it or not:

```tsx
// plugins/your-plugin/src/plugin.ts
import { createPlugin } from '@backstage/core-plugin-api';
import { createSearchResultListItemExtension } from '@backstage/plugin-search-react';

const plugin = createPlugin({ id: 'YOUR_PLUGIN_ID' });

export const YourSearchResultListItem = plugin.provide(
  createSearchResultListItemExtension({
    name: 'YourSearchResultListItem',
    component: () => import('./components').then(m => m.YourSearchResultListItem),
    // Only results matching your type will be rendered by this extension
    predicate: result => result.type === 'YOUR_RESULT_TYPE', ⬅️ 
  }),
);
```

Now that you know how a search result item is provided, let's finally see how they can be used, for example, to compose a page in your application:

```tsx
// packages/app/src/components/search/searchPage.ts
import React from 'react';

import { Grid, Paper } from '@material-ui/core';
import BuildIcon from '@material-ui/icons/Build';

import {
  Page,
  Header,
  Content,
  DocsIcon,
  CatalogIcon,
} from '@backstage/core-components';
import { SearchBar, SearchResult } from '@backstage/plugin-search-react';

// Your search result item extension
import { YourSearchResultListItem } from '@backstage/your-plugin';

// Extensions provided by other plugin developers
import { ToolSearchResultListItem } from '@backstage/plugin-explore';
import { TechDocsSearchResultListItem } from '@backstage/plugin-techdocs';
import { CatalogSearchResultListItem } from '@internal/plugin-catalog-customized';

// This example omits other components, like filter and pagination
const SearchPage = () => (
  <Page themeId="home">
    <Header title="Search" />
    <Content>
      <Grid container direction="row">
        <Grid item xs={12}>
          <Paper>
            <SearchBar />
          </Paper>
        </Grid>
        <Grid item xs={12}>
          <SearchResult>
            <YourSearchResultListItem /> ⬅️ 
            <CatalogSearchResultListItem icon={<CatalogIcon />} /> ⬅️ 
            <TechDocsSearchResultListItem icon={<DocsIcon />} /> ⬅️ 
            <ToolSearchResultListItem icon={<BuildIcon />} /> ⬅️ 
          </SearchResult>
        </Grid>
      </Grid>
    </Content>
  </Page>
);
export const searchPage = <SearchPage />;
```

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/app/src/components/search/SearchModal.tsx
Comment (aitherios) [inline packages/app/src/components/search/SearchModal.tsx:207]: This is way better to read =D

File: packages/core-components/package.json
Comment (aitherios) [inline packages/core-components/package.json:70]: Was this necessary?

Comment (aitherios) [reply packages/core-components/package.json:70]: I didn't see any other file in core-components being changed, just this package added.

File: plugins/rollbar-backend/src/api/RollbarApi.ts
Comment (aitherios) [inline plugins/rollbar-backend/src/api/RollbarApi.ts:107]: I assume there probably was a type mismatch, can you hint me on why you had to change this (and the next function getForProject)?

File: plugins/techdocs-react/package.json
Comment (aitherios) [inline plugins/techdocs-react/package.json:45]: This is here because of the `techdocs` package?

**Review by aitherios — APPROVED** [review]
Approving with some comments but nothing blocking a merge.

File: plugins/techdocs-react/package.json
Comment (camilaibs) [reply plugins/techdocs-react/package.json:45]: @aitherios see [here](https://github.com/backstage/backstage/pull/15855/files#diff-a768fbf60c0ae56ed8e72f6d5c6d9d5ff24fdc03484f0f8dadf164a13f04c2a5), the error mentioned in the changeset happens when I try to update the API reports for this package, and I need to update the report because I'm exporting a new extension. @backstage/techdocs-core do you have any recommendations?

File: plugins/rollbar-backend/src/api/RollbarApi.ts
Comment (camilaibs) [reply plugins/rollbar-backend/src/api/RollbarApi.ts:107]: @aitherios I suggested this fix because the build fails when running `tsc:full`. The generic type `T` is passed down to the `camelcaseKeys` function, and it requires a type that extends a array or record, see the message below:
image
@backstage/maintainers do you have any recommendations on how to fix the build error?

File: packages/core-components/package.json
Comment (camilaibs) [reply packages/core-components/package.json]: @aitherios it was a tentative of fixing this build error, but I removed this since it does not fix the problem:
https://github.com/backstage/backstage/actions/runs/3985180683/jobs/6832242810
@backstage/maintainers do you have any recommendation?

Comment (Rugvip) [reply packages/core-components/package.json]: It's a breakage in a transitive upstream dependency. Looks like they're on it though so hopefully fixed soon https://github.com/syntax-tree/unist-util-position/issues/10

Comment (camilaibs) [reply packages/core-components/package.json]: @Rugvip How does this fix sound to you?
https://github.com/backstage/backstage/pull/15855/commits/136e333f6ce77e3993a751e3e864447f361175a1

File: plugins/techdocs-react/package.json
Comment (camilaibs) [reply plugins/techdocs-react/package.json:45]: @Rugvip also seems to be a breakage in a transitive upstream dependency.

File: packages/core-components/package.json
Comment (camilaibs) [reply packages/core-components/package.json]: Last commit removed, it is working because a new release is out (Thanks @Rugvip)

File: plugins/rollbar-backend/src/api/RollbarApi.ts
Comment (camilaibs) [reply plugins/rollbar-backend/src/api/RollbarApi.ts:107]: @Rugvip also seems to be a breakage in a transitive upstream dependency.

File: plugins/techdocs-react/package.json
Comment (camilaibs) [reply plugins/techdocs-react/package.json]: It seems to be fixed, closing this.

**Review by agentbellnorm — APPROVED** [review]
Nice, great improvement!

File: plugins/catalog/src/plugin.ts
Comment (Rugvip) [inline plugins/catalog/src/plugin.ts:262]: Would it make sense to deprecate the old component?

File: plugins/rollbar-backend/src/api/RollbarApi.ts
Comment (Rugvip) [reply plugins/rollbar-backend/src/api/RollbarApi.ts:107]: I'd say it's best to move this fix to a separate PR if it's still needed. We had a couple of upstream breakages the last week but they should all be patched by now.

In general it's best to keep these kind of fixes in their own PR since more people are likely affected and would want to ship it faster.

File: plugins/search-react/api-report.md
Comment (Rugvip) [inline plugins/search-react/api-report.md:41]: Please use more verbose type parameter names, in this case In guessing it's `TProps`? More on that in the [style guide](https://github.com/backstage/backstage/blob/master/STYLE.md#naming)

File: plugins/search-react/src/extensions.tsx
Comment (Rugvip) [inline plugins/search-react/src/extensions.tsx:56]: I recommend using `useElementFilter` here, it helps you filter through children and find component data,  etc. It also takes care of iterating through nested arrays and fragments as well as support for feature flags

File: plugins/search-react/src/extensions.tsx
Comment (Rugvip) [inline plugins/search-react/src/extensions.tsx:114]: How come you didn't end up supporting lazy loaded components here? Are you expecting these components to always be very small?

**Review by Rugvip — CHANGES_REQUESTED** [review]
Nice 😁 makes the search page a lot cleaner

Couple of style nits and whatnot

File: plugins/search-react/api-report.md
Comment (camilaibs) [reply plugins/search-react/api-report.md]: Done 🙌🏻

File: plugins/search-react/src/extensions.tsx
Comment (camilaibs) [reply plugins/search-react/src/extensions.tsx:52]: The elements were filtered beforehand using `useElementFilter` [here](https://github.com/backstage/backstage/pull/15855/files#diff-4c59e19e90caafd5ccb90cb3236950aa33e04d44d061b4b98aa4cbd3bda8f295R160). In this function I just wanted to get their predicate and component from data. Is there a way to get elements and data all at once?

File: plugins/search-react/src/extensions.tsx
Comment (camilaibs) [reply plugins/search-react/src/extensions.tsx]: We're passing a component that always returns null to createReactExtension [here](https://github.com/backstage/backstage/pull/15855/files#diff-4c59e19e90caafd5ccb90cb3236950aa33e04d44d061b4b98aa4cbd3bda8f295R132), would you recommend using `lazy` for this case?

File: plugins/catalog/src/plugin.ts
Comment (camilaibs) [reply plugins/catalog/src/plugin.ts:262]: I'm not sure if it makes sense to deprecate because the extensions only return null if they are used just as simple components. It seems that returning null isn't correct if we want to replace the old component...

File: plugins/search-react/src/extensions.tsx
Comment (Rugvip) [reply plugins/search-react/src/extensions.tsx]: No but you could make the the component itself that is registered in the component data be lazy loaded. Not saying that it's necessarily better, scaffolder field extensions are sync for example. Just wanna make sure it's considered. I don't know myself if search items are large enough for it to make sense. It can also be added later on by adding more options, so it's mostly for the initial implementation and the existing extensions.

File: plugins/search-react/src/extensions.tsx
Comment (Rugvip) [reply plugins/search-react/src/extensions.tsx:52]: You can select multiple things at use in the filter function, but since you need access to the props as well that's probably just gonna make things messy. I'd say keep things as is 👍

File: plugins/catalog/src/plugin.ts
Comment (Rugvip) [reply plugins/catalog/src/plugin.ts:262]: It would be the complete migration package. Deprecate the old components and ask users to move over to this new pattern

Comment (camilaibs) [reply plugins/catalog/src/plugin.ts]: Thanks for commenting on deprecation, I changed the extension implementation to be compatible with the old components (instead of returning null, the extensions now render the original component), and I exported the extensions in place of the old components; this way, it is transparent to end consumers, they will automatically use the extensions behind the scenes when importing the old component. Sounds good?

File: plugins/search-react/src/extensions.tsx
Comment (camilaibs) [reply plugins/search-react/src/extensions.tsx]: The components are list items and can be as simple as the default component that we export as having a preview and querying external sources as our internal use case, so I changed it to lazy components; thanks for bringing that up.

File: plugins/catalog/src/plugin.ts
Comment (Rugvip) [reply plugins/catalog/src/plugin.ts]: Oh interesting, yep I think that makes sense! 👍

**Review by Rugvip — APPROVED** [review]
👌

**Pull request merged**
